### PR TITLE
net: l2: ieee802154 improved association request and response procedure (incl. test coverage)

### PR DESCRIPTION
**Pull request by ghost**
This is the next change set in the #50336 series.

First a few minor changes are introduced that probably do not justify their own PR and should be easy to review although they are somewhat unrelated. Please let me know if you want to see them in a separate PR.

The main focus of this PR is on the association request and response procedure:
* The corresponding shell and net_mgmt command is covered by a new integration test.
* This test is rather involved (it needs to verify/mock 4 packages: an association request, then a mocked ACK and a mocked response which is again acknowledged).
* As the IEEE 802.15.4 stack does not provide hooks to inject response packets from test code, I rely on placing responses into the RX queue before sending the TX packet as the RX packets will only be processed by L2 once the TX path yields which will place them into the RX path just at the right time.

It turned out that the associate and disassociate commands probably never really worked as they had some fundamental flaws (see the commit comments). I tried hard to make several changes to make the fixes easier to review but I didn't succeed w/o breaking the build. If you have an idea how to break this up let me know. But then probably it is not so important even if there are still some bugs in there as it cannot be more broken than before. ;-)

Disassociation does not yet have test coverage - but then again - it cannot be more broken now than it already was. And the PR is already large enough. Test coverage for disassociation will follow in a later PR.

Finally I tried to simplify the locking scheme of the ACK procedure. To be honest, I'm not even sure that ACK properly worked before - although the tests seemed to indicate that it did. I tested this simplified procedure on hardware but it would be great if you had a closer look whether this kind of locking really worked and maybe also why it worked before (or not)...

Anyway let me know if you have any ideas how I can make this easier to review.

### CODE REVIEW

**Comment by fgrandel** [comment]
UPDATE: Seems that some tests break in the CI/CD pipeline while they don't break locally. Maybe the test strategy doesn't work so well after all. Will have to look into this tomorrow.

**Review by cfriedt — DISMISSED** [review]
LGTM - just some comments.

In general, it would be nice if eventually we could separate getting / setting "radio" attributes into MIB and PIB queries to be closer to spec terminology.

**Comment by cfriedt** [comment]
Just curious if you've tested on multiple hw platforms.

File: include/zephyr/net/ieee802154.h
Comment (fgrandel) [reply include/zephyr/net/ieee802154.h:159]: @cfriedt You're right - the way it looks now the bitfield could be replaced by a simple boolean. I'm also not a big fan of bitfields as they are underspecified and caused famous portability and security issues in the past. I do have commits in my dev branch that will use that bitfield for various flags and I'm not aware of an alternative that wouldn't require more RAM. The _unused is just to remind me of the number of bits I have available to not introduce an additional byte, but definitely not required and can be removed in the final version. Makes somehow sense?

**Comment by fgrandel** [comment]
> Just curious if you've tested on multiple hw platforms.

No - I'm testing this on CC13xx chips only. But most of the code I'm touching in the native stack was so badly broken before anyway that I'm quite sure that no one used that so far. We'll test quite intensively on various hardware once we have the TSCH stack running. We'll also do compat tests with Contiki then - one of the maintainers, Atis Elst promised to help us with that.

**Comment by fgrandel** [comment]
> it would be nice if eventually we could separate getting / setting "radio" attributes into MIB and PIB queries to be closer to spec terminology.

@cfriedt Yes, this is true. As a first step I'm documenting the correspondence of standard primitives and native code features wherever possible, this is ongoing work. The long term goal is: https://github.com/zephyrproject-rtos/zephyr/pull/58438#discussion_r1232470448 (which includes your proposal but is more general).

**Review by cfriedt — DISMISSED** [review]

**Comment by fgrandel** [comment]
UPDATE: Just wrote tests for setting the external address and disassociation via shell (not part of this PR to not overload it), confirming that both work. The tests will be provided in a separate PR once this one is merged.

**Comment by fgrandel** [comment]
@jukkar @nandojve heads up to you. This time it's about association/disassociation shell cmd/net_mgmt fixes in the native ieee802154 L2 in case you want to have a look.

**Review by cfriedt — DISMISSED** [review]

**Review by rlubos — APPROVED** [review]
LGTM, thanks

**Review by cfriedt — APPROVED** [review]

**Pull request merged**
